ext/atomic/index.md

```diff
@@ -0,0 +1,510 @@
+# Atomic Operations Extension
+
+This extension provides a means to perform multiple "operations" in a linear and
+atomic manner. Operations are a serialized form of the mutations allowed in the
+base JSON:API specification.
+
+Clients can send an array of operations in a single request. This extension
+guarantees that those operations will be processed in order and will either
+completely succeed or fail together.
+
+## <a href="#namespace" id="namespace" class="headerlink"></a> Namespace
+
+This extension uses the namespace `atomic`.
+
+> Note: JSON:API extensions can only introduce new document members using a
+> reserved namespace as a prefix.
+
+## <a href="#document-structure" id="document-structure" class="headerlink"></a> Document Structure
+
+A document that supports this extension **MAY** include any of the top-level
+members allowed by the base specification with the exception of `data` and
+`included`, which **MUST NOT** be included.
+
+In addition, such a document **MAY** include either of the following members,
+but not both:
+
+- `atomic:operations` - an array of [operation objects](#operation-objects).
+
+- `atomic:results` - an array of [result objects](#result-objects).
+
+If either `atomic:operations` or `atomic:results` is present, the `errors`
+member **MUST NOT** be included in the same document.
```

Comment:
Can this somehow be relaxed? I'm thinking about streaming large payloads. It would be convenient to process operation as they are streamed and stream back the results until an error occurs at which point the server would revert all operations and send an error. Potentially even continuing processing to provide to the client a full error list for the entire payload. So rather than having `errors` as a top-level element only, how about allowing an `error` as a result type?

Reply:
I think that that use case is sufficiently on the edge that we should not alter the design of this extension to support it. Turning errors into a result type would probably be confusing for the majority of users who will not understand why the spec allows individual errors, yet forces the entire request to roll back. For example, many people would probably want to use this spec for bulk creation + retries if that were the case.

All that said, I _do_ really see value in a different extension that better suits your needs.